### PR TITLE
[autoscaler][kubernetes] Helm chart

### LINKED ISSUE #12654: [k8s] [autoscaler] Helm chart for deploying Ray on Kubernetes
### Describe your feature request
This is a commonly requested feature, and it's on our to-do list.

For feature requests or questions, post on our Discussion page instead: https://discuss.ray.io/

### LINKED ISSUE #15652: [Ray Operator] The ServiceAccount should not be hardcoded to the 'default' namespace
### What is the problem?

When deploying the cluster-scoped Ray Operator, the ServiceAccount `ray-operator-serviceaccount` is hardcoded to be created in the `default` namespace, and the corresponding ClusterRoleBinding also binds that SA in the `default` namespace. This prevents Ray Operator from being deployed in any other namespace, because there is no such SA in the target namespace.

### Reproduction (REQUIRED)

- [x] I have verified my script runs in a clean environment and reproduces the issue.
- [x] (NA) I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

```bash
$ kubectl set-context --current --namespace ray

$ kubectl apply -f https://raw.githubusercontent.com/ray-project/ray/master/python/ray/autoscaler/kubernetes/operator_configs/operator_cluster_scoped.yaml
serviceaccount/ray-operator-serviceaccount created
clusterrole.rbac.authorization.k8s.io/ray-operator-clusterrole created
clusterrolebinding.rbac.authorization.k8s.io/ray-operator-clusterrolebinding created
deployment.apps/ray-operator created

$ kubectl get sa
NAME      SECRETS   AGE
default   1         6d21h

$ kubectl get sa -n default
NAME                          SECRETS   AGE
default                       1         97d
ray-operator-serviceaccount   1         5m55s

$ kubectl get po
No resources found in ray namespace.

$ kubectl get deployments
NAME           READY   UP-TO-DATE   AVAILABLE   AGE
ray-operator   0/1     0            0           7m1s

$ kubectl get rs
NAME                      DESIRED   CURRENT   READY   AGE
ray-operator-5d896d64c5   1         0         0       7m27s

$ kubectl describe rs ray-operator-5d896d64c5
Name:           ray-operator-5d896d64c5
Namespace:      ray
Selector:       cluster.ray.io/component=operator,pod-template-hash=5d896d64c5
Labels:         cluster.ray.io/component=operator
                pod-template-hash=5d896d64c5
Annotations:    deployment.kubernetes.io/desired-replicas: 1
                deployment.kubernetes.io/max-replicas: 2
                deployment.kubernetes.io/revision: 1
Controlled By:  Deployment/ray-operator
Replicas:       0 current / 1 desired
Pods Status:    0 Running / 0 Waiting / 0 Succeeded / 0 Failed
Pod Template:
  Labels:           cluster.ray.io/component=operator
                    pod-template-hash=5d896d64c5
  Service Account:  ray-operator-serviceaccount
  Containers:
   ray:
    Image:      rayproject/ray:nightly
    Port:       <none>
    Host Port:  <none>
    Command:
      ray-operator
    Limits:
      memory:  2Gi
    Requests:
      cpu:        1
      memory:     1Gi
    Environment:  <none>
    Mounts:       <none>
  Volumes:        <none>
Conditions:
  Type             Status  Reason
  ----             ------  ------
  ReplicaFailure   True    FailedCreate
Events:
  Type     Reason        Age                 From                   Message
  ----     ------        ----                ----                   -------
  Warning  FailedCreate  38s (x14 over 79s)  replicaset-controller  Error creating: pods "ray-operator-5d896d64c5-" is forbidden: error looking up service account ray/ray-operator-serviceaccount: serviceaccount "ray-operator-serviceaccount" not found

```

The `ClusterRoleBinding` does require its `subject`s to specify a namespace, so this isn't easy to automate for all use cases out-of-the box, but perhaps at least the documentation at https://docs.ray.io/en/master/cluster/kubernetes.html#picking-a-kubernetes-namespace can be updated to avoid end-user surprises, until a Helm- or Kustomize-based deployment is available.

### PR DESCRIPTION
**Pull request by DmitriGekhtman**
## Why are these changes needed?

This PR adds a Helm chart to simplify deployment of the Ray operator.
For now, only basic configuration options are exposed -- we could later expose more configuration on user feedback.

UX:
Launching operator and cluster with default settings. 
`helm -n <namespace> install <release name> ./ray`
For custom configuration values one, uses `-f <custom-values.yaml>` or `--set <option>=value`.
Can launch just operator, without Ray cluster using `--set operatorOnly=true`.
Can create a RayCluster resource without the operator using `--set clusterOnly=true`:
if the operator is already running, this is the way to create multiple ray clusters.

This PR also pins Ray images used in the Ray clusters to 1.3.0
The operator is still set to nightly for now, for better stability. There's comment in the helm chart noting this and the alternatives of using 1.3.0 or a specific Ray master commit.

## Related issue number

Closes #12654
Mostly addresses #15222 (still using nightly operator image, for now)
Closes #15652

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by DmitriGekhtman** [comment]
Before adding documentation, I could use some feedback on whether the configuration options exposed in `values.yaml` look sensible.
cc @tgaddair @ijrsvt

**Review by ijrsvt — COMMENTED** [review]
@DmitriGekhtman Is this file auto-generated or did you write this by hand: 
python/ray/autoscaler/kubernetes/operator_configs/helm/ray/crds/cluster_crd.yaml

**Comment by yiranwang52** [comment]
nice!

**Comment by edoakes** [comment]
@DmitriGekhtman the config options look very reasonable to me as a first cut!

File: python/ray/autoscaler/kubernetes/operator_configs/helm/ray/values.yaml [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    maxWorkers: 3
+    # Memory used by this Pod type.
+    # (Used for both requests and limits.)
+    memory: 512Mi
+    # Number of CPUs used by this pod type.
+    # (Used for both requests and limits. Must be an integer, as Ray does not support fractional CPUs.)
+    numCPU: 1
+    # Number of NVIDIA GPUs used by this pod type.
+    # (Optional, requires GPU nodes with appropriate setup. See https://docs.ray.io/en/master/cluster/kubernetes-gpu.html)
+    numGPU: 0
```
Comment (tgaddair) [inline python/ray/autoscaler/kubernetes/operator_configs/helm/ray/values.yaml:38]: You may also want to allow the user to specify the GPU sku here as well. For example:

```
nodeSelector:
    cloud.google.com/gke-accelerator: nvidia-tesla-v100
```

See https://cloud.google.com/kubernetes-engine/docs/how-to/gpus#multiple_gpus

File: python/ray/autoscaler/kubernetes/operator_configs/helm/ray/templates/raycluster.yaml [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+            imagePullPolicy: Always
+            image: {{ $.Values.image }}
+            # Do not change this command - it keeps the pod alive until it is
+            # explicitly killed.
+            command: ["/bin/bash", "-c", "--"]
+            args: ['trap : TERM INT; sleep infinity & wait;']
+            ports:
+            - containerPort: 6379  # Redis port
+            - containerPort: 10001  # Used by Ray Client
+            - containerPort: 8265  # Used by Ray Dashboard
```
Comment (tgaddair) [inline python/ray/autoscaler/kubernetes/operator_configs/helm/ray/templates/raycluster.yaml:49]: May be useful to include the Ray Serve port on 8000 as well.

**Review by tgaddair — APPROVED** [review]
Nice! Just a couple suggestions.

**Comment by DmitriGekhtman** [comment]
> @DmitriGekhtman Is this file auto-generated or did you write this by hand:
> python/ray/autoscaler/kubernetes/operator_configs/helm/ray/crds/cluster_crd.yaml

This was copied from  `kubernetes/operator_configs/cluster_crd.yaml`.
That file has the schema for a Kubernetes podSpec _pasted in_ which explains the length.
If you fold [podConfig.spec](https://github.com/ray-project/ray/blob/6570474c344c9a2a510737fdc07dccbeeb6f76bf/python/ray/autoscaler/kubernetes/operator_configs/cluster_crd.yaml#L95) the file looks much smaller.
It's an open item to do podSpec validation in a saner way:
https://github.com/ray-project/ray/issues/13674

File: python/ray/autoscaler/kubernetes/operator_configs/helm/ray/templates/raycluster.yaml (hunk 2 above)
Comment (DmitriGekhtman) [reply python/ray/autoscaler/kubernetes/operator_configs/helm/ray/templates/raycluster.yaml:49]: From what I understand, a container's `ports` list functions pretty much as documentation https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/pod-v1/#ports 
That said -- it's definitely a good idea to document the serve port here!

**Comment by DmitriGekhtman** [comment]
I've added the `nodeSelector` field.
I've also exposed the `rayResources` field, so that users can signal the presence of custom resources (e.g. gpu type).

Now will add some tests and documentation.

**Comment by DmitriGekhtman** [comment]
Not sure about the ideal protocol for versioning and distributing this chart.
Will solve that later -- for now it will be available by cloning Ray master.

**Comment by DmitriGekhtman** [comment]
docs written --  this should be good for a final review.
@richardliaw @javi-redondo could you guys glance at the docs?

**Comment by DmitriGekhtman** [comment]
Docs generated from Build here:
https://ray--15614.org.readthedocs.build/en/15614/cluster/kubernetes.html

**Review by ijrsvt — APPROVED** [review]
This looks good to me!

**Comment by DmitriGekhtman** [comment]
@ijrsvt It's good to merge
-- main thing was to get +1s on`values.yaml`

**Pull request merged**